### DISCUSSION TITLE
Explanation and question about "Type-safe routes" and if there are plans to implement them

### DISCUSSION THREAD

**Discussion by jacob-emery** [csells/go_router#159] [Q&A] [answered]
I was reading a `go_router` announcement on reddit by Chris Sells, and in one of the comments he wrote: _"If you like auto_route and it serves your needs, then I'd recommend that you keep using it. Personally, I didn't like the builder requirement, which I think adds to the complexity, but I like that auto_route has typesafe routes."_.
Since I didn't know what type-safe routes are. I googled them but I could only find Remi Rousselet's explanation (_"The benefit of Navigator 2.0 isn't being declarative. Instead, it's a type-safe way to push routes since we don't push URLs anymore but rather update our state. That is a noble goal, but not the same as the one advertised"_). I still don't understand the concept very well. Could anyone here explain this concept more in detail, and in very simple terms? Maybe this information could also be added to the docs for other people to choose if `go_router`is the right package for them.

Besides the above question, I have 5 more question about type-safe routes..
1. Is this a limitation because of an absence of a code generator, such as `auto_router_generator`?
2. Is it a feature that can be added in the future?
3. When are routes created with go_router? Are they generated at compile time or runtime? I am wondering because you don't need a generator.
4. Is there any more detail we can get about the design of the package and some arguments why type-safe routes are not a priority?
5. In which cases would type-safe routes be an objective must have? 

Disclaimer: This question is not only about `go_router`. Also, I am junior developer, therefore I am aware that some of the concepts might seem obvious for the ones reading my question. But I would still like the get a clear understanding :)

**Comment by csells** [comment, marked as answer]
Type safety is the idea that the compiler catches your mistakes instead of waiting until run-time. For example, if you have the following route:

```dart
GoRouter(path: '/family/:fid', ...)
```

and navigate to it w/o the 'fid' parameter:

```dart
context.go('/family'); // oops. forgot the fid param
```

this code will compile just fine. You won't see the exception until run-time because you should've called it like so:

```dart
context.go('/family', params: {'fid': 'f2'});
```

However, if this was a typesafe route, you could imagine calling it like so:

```dart
familyRoute.go(context, fid: 'f2');
```

In this case, the compiler would let you know if you forgot to add the `fid` parameter.

The idea of using a code generator in Dart is to be able to take some data, like the list of routes, and turn that into a set of types w/o having to write that code yourself. The auto_route_generator takes the definition of routes and provides the types to enable typesafe routing.

Obviously you don't need type safe routes to enable routing (since go_router has lots of happy customers), but it would be nice to have as an optional add-on. We've been working on such a thing for go_router, but we don't have it yet.

**Reply by jacob-emery** [reply]
This answer thoroughly explains type-safe routes. Thank you a lot!
